src/riak_kv_wm_timeseries.erl

```diff
@@ -29,572 +34,375 @@
-                FalseWithDetails ->
-                    FalseWithDetails
-            end;
-        FalseWithDetails ->
-            FalseWithDetails
+-spec content_types_accepted(#wm_reqdata{}, #ctx{}) -> cb_rv_spec([{string(), atom()}]).
+content_types_accepted(RD, Ctx) ->
+    content_types_accepted(wrq:path_tokens(RD), RD, Ctx).
+
+content_types_accepted([], RD, Ctx) ->
+    %% the JSON in the POST will be handled by process_post,
+    %% so this handler will never be called.
+    {[{"application/json", undefined}], RD, Ctx};
+content_types_accepted(_, RD, Ctx) ->
+    {[], RD, Ctx}.
+
+-spec resource_exists(#wm_reqdata{}, #ctx{}) -> cb_rv_spec(boolean() | halt()).
+resource_exists(RD, #ctx{mod=Mod} = Ctx) ->
+    case riak_kv_wm_ts_util:table_module_exists(Mod) of
+        true ->
+            resource_exists(wrq:path_tokens(RD), wrq:method(RD), RD, Ctx);
+        false ->
+            Resp = riak_kv_wm_ts_util:set_error_message("table ~p not created", [Mod], RD),
+            {false, Resp, Ctx}
     end.
 
--spec validate_request(#wm_reqdata{}, #ctx{}) -> ?CB_RV_SPEC.
-validate_request(RD, Ctx) ->
-    case wrq:path_info(api_version, RD) of
-        "v1" ->
-            validate_request_v1(RD, Ctx);
-        BadVersion ->
-            handle_error({unsupported_version, BadVersion}, RD, Ctx)
+resource_exists([], 'POST', RD, Ctx) ->
+    {true, RD, Ctx};
+resource_exists(Path, 'GET', RD,
+                #ctx{table=Table,
+                     mod=Mod,
+                     options=Options}=Ctx) ->
+    %% Would be nice if something cheaper than using get_data existed to check
+    %% if a key is present.
+    try
+        Key = validate_key(Path, Mod),
+        case riak_kv_ts_api:get_data(Key, Table, Mod, Options) of
+            {ok, Record} ->
+                {true, RD, Ctx#ctx{object=Record,
+                                   key=Key}};
+            {error, notfound} ->
+                {{halt, 404}, RD, Ctx};
+            {error, InternalReason} ->
+                InternalResp = riak_kv_wm_ts_util:set_error_message("Internal error: ~p", [InternalReason], RD),
+                {{halt, 500}, InternalResp, Ctx}
+        end
+    catch
+        _:Reason ->
+            Resp = riak_kv_wm_ts_util:set_error_message("lookup on ~p failed due to ~p",
+                                                        [Path, Reason],
+                                                        RD),
+            {false, Resp, Ctx}
+    end;
+resource_exists(Path, 'DELETE', RD, #ctx{mod=Mod}=Ctx) ->
+    %% Since reading the object is expensive we will assume for now that the
+    %% object exists for a delete, but if it turns out that it does not then the
+    %% processing of the delete will return 404 at that point.
+    try
+        Key = validate_key(Path, Mod),
+        {true, RD, Ctx#ctx{key=Key}}
+    catch
+        _:Reason ->
+            Resp = riak_kv_wm_ts_util:set_error_message("lookup on ~p failed due to ~p",
```

Comment:
Using `handle_error` here (and elsewhere in this module) would be logical, instead of custom setting of response body.